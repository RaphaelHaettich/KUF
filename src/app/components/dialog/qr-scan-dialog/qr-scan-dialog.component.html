<div mat-dialog-content class="{{ componentName }}__header">
    <h2 mat-dialog-title class="{{ componentName }}__title">{{ data.title }}</h2>
</div>
<ob-spinner [channel]="spinnerChannel" />
<div mat-dialog-content class="{{ componentName }}__content">
    <ngx-scanner-qrcode #action="scanner"></ngx-scanner-qrcode>
</div>
<div mat-dialog-actions class="{{ componentName }}__actions">
    <button (click)="closeDialog()" type="button" mat-button obButton="primary" mat-dialog-close>
        {{ "i18n.common.close" | translate }}
    </button>
</div>
