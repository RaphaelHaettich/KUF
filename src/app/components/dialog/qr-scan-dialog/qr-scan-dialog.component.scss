@import '@oblique/oblique/styles/scss/core/mixins/typography';
@import '@oblique/oblique/styles/scss/core/mixins/layout';
@import '@oblique/oblique/styles/scss/core/variables';

.qr-scan-dialog {
    &__title {
        font-size: $ob-font-size-xl !important;
    }

    &__actions,
    &__header {
        display: flex;
        justify-content: center;
    }
}
