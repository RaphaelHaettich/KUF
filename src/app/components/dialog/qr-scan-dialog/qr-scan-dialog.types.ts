export type QrScanData = {
    qrPrefix: string;
    title: string;
    callback: (result: string) => unknown;
};
