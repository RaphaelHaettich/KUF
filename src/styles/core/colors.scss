$kuf-red: #dc0018;
$kuf-error: #C70023;
$kuf-yellow: #FF9B00;
$kuf-green: #00813a;
$kuf-grey: #808080;
$kuf-lightgrey: #E5E5E5;
$kuf-blue: #006699;
