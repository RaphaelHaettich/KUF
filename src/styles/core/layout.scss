#content {
    padding: 0.3rem !important;
    background-color: rgba(216, 216, 216, 0.603);
}
