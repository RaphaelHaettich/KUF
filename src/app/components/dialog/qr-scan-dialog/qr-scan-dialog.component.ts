import {AfterViewInit, Component, OnDestroy, ViewChild, inject} from '@angular/core';
import {MAT_DIALOG_DATA, MatDialog} from '@angular/material/dialog';
import {NgxScannerQrcodeComponent, ScannerQRCodeDevice} from 'ngx-scanner-qrcode';
import {QrScanData} from './qr-scan-dialog.types';
import {SubSink} from 'subsink';
import {ObNotificationService, ObSpinnerService} from '@oblique/oblique';

@Component({
    selector: 'app-qr-scan-dialog',
    templateUrl: './qr-scan-dialog.component.html',
    styleUrl: './qr-scan-dialog.component.scss',
})
export class QrScanDialogComponent implements AfterViewInit, OnDestroy {
    readonly componentName = 'qr-scan-dialog';
    readonly spinnerChannel = 'qrScanDialogSpinner';
    readonly data = inject<QrScanData>(MAT_DIALOG_DATA);
    readonly spinnerService = inject(ObSpinnerService);
    readonly obNotificationService = inject(ObNotificationService);
    readonly dialog = inject(MatDialog);
    @ViewChild('action') action!: NgxScannerQrcodeComponent;
    private readonly subSink = new SubSink();

    ngOnDestroy() {
        this.subSink.unsubscribe();
        this.spinnerService.forceDeactivate(this.spinnerChannel);
    }

    ngAfterViewInit(): void {
        this.spinnerService.activate(this.spinnerChannel);
        this.action.isReady.subscribe(() => {
            const playDeviceFacingBack = (devices: ScannerQRCodeDevice[]) => {
                const device = devices.find((device: ScannerQRCodeDevice) =>
                    /back|rear|environment/gi.test(device.label),
                );

                this.action.playDevice(device ? device.deviceId : devices[0].deviceId);
            };
            this.subSink.sink = this.action.start(playDeviceFacingBack).subscribe(
                () => {
                    this.spinnerService.deactivate(this.spinnerChannel);
                },
                () => {
                    this.obNotificationService.error({
                        message: 'Kamera konnte gestartet werden',
                    });
                    this.spinnerService.deactivate(this.spinnerChannel);
                },
            );
            this.subSink.sink = this.action.data.subscribe((result) => {
                const qrValue = result?.[0]?.value;
                if (!qrValue || !qrValue.startsWith(this.data.qrPrefix)) {
                    return;
                }
                this.action.stop();
                this.data.callback(qrValue.replace(this.data.qrPrefix, ''));
            });
        });
    }

    public closeDialog() {
        this.action.stop();
        this.dialog.closeAll();
    }
}
