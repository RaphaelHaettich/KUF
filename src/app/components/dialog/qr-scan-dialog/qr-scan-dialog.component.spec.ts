import {ComponentFixture, TestBed} from '@angular/core/testing';
import {MAT_DIALOG_DATA, MatDialog} from '@angular/material/dialog';
import {QrScanDialogComponent} from './qr-scan-dialog.component';
import {HttpClientModule} from '@angular/common/http';
import {TranslateModule} from '@ngx-translate/core';
import {ObSpinnerModule, multiTranslateLoader} from '@oblique/oblique';
import {NgxScannerQrcodeComponent, ScannerQRCodeResult} from 'ngx-scanner-qrcode';
import {BrowserAnimationsModule} from '@angular/platform-browser/animations';
import {MockBuilder} from 'ng-mocks';
import {AsyncSubject} from 'rxjs';

describe('QrScanDialogComponent', () => {
    let component: QrScanDialogComponent;
    let fixture: ComponentFixture<QrScanDialogComponent>;
    const callback = jest.fn();
    const qrPrefix = 'qrPrefix';

    beforeEach(async () => {
        await TestBed.configureTestingModule({
            declarations: [QrScanDialogComponent, NgxScannerQrcodeComponent],
            imports: [
                TranslateModule.forRoot(multiTranslateLoader()),
                HttpClientModule,
                ObSpinnerModule,
                BrowserAnimationsModule,
            ],
            providers: [
                MatDialog,
                {
                    provide: MAT_DIALOG_DATA,
                    useValue: {
                        qrPrefix,
                        title: 'title',
                        callback,
                    },
                },
            ],
        }).compileComponents();
    });

    beforeEach(() => MockBuilder(QrScanDialogComponent, class {}));

    beforeEach(() => {
        jest.clearAllMocks();
        fixture = TestBed.createComponent(QrScanDialogComponent);
        component = fixture.componentInstance;
        fixture.detectChanges();
    });

    it('should test the scanning and component rendering correctly', () => {
        const isReadySubscribeSpy = jest.spyOn(component.action.isReady, 'subscribe');
        const stopSpy = jest.spyOn(component.action, 'stop');
        const dataSpy = jest.spyOn(component.action.data, 'subscribe');
        const spinnerDeactivateSpy = jest.spyOn(component.spinnerService, 'deactivate');
        const errorNotificationSpy = jest.spyOn(component.obNotificationService, 'error');
        const playDeviceSpy = jest
            .spyOn(component.action, 'playDevice')
            .mockReturnValue(new AsyncSubject<null>());
        const startSpy = jest
            .spyOn(component.action, 'start')
            .mockReturnValue(new AsyncSubject<null>());
        const startSubscribeSpy = jest.spyOn(component.action.start(), 'subscribe');

        component.ngAfterViewInit();

        isReadySubscribeSpy.mock.calls?.[0][0]?.(true);
        startSpy.mock.calls?.[1][0]?.([{deviceId: 'deviceId', label: 'backlabel'}]);
        startSubscribeSpy.mock.calls?.[0][0]?.(true);
        startSubscribeSpy.mock.calls?.[0][1]?.(true);
        dataSpy.mock.calls?.[0][0]?.([{value: `${qrPrefix}/url`} as ScannerQRCodeResult]);

        expect(playDeviceSpy.mock.calls).toMatchSnapshot();
        expect(spinnerDeactivateSpy.mock.calls).toMatchSnapshot();
        expect(errorNotificationSpy.mock.calls).toMatchSnapshot();
        expect(callback.mock.calls).toMatchSnapshot();
        expect(stopSpy).toHaveBeenCalled();
        expect(fixture).toMatchSnapshot();
    });

    it('should stop the action and close the dialog on closeDialog()', () => {
        const actionSpy = jest.spyOn(component.action, 'stop');
        const dialogSpy = jest.spyOn(component.dialog, 'closeAll');

        component.closeDialog();

        expect(actionSpy).toHaveBeenCalled();
        expect(dialogSpy).toHaveBeenCalled();
    });
});
