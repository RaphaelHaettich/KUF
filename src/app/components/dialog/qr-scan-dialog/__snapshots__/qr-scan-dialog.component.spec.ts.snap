// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QrScanDialogComponent should test the scanning and component rendering correctly 1`] = `
[
  [
    "deviceId",
  ],
]
`;

exports[`QrScanDialogComponent should test the scanning and component rendering correctly 2`] = `
[
  [
    "qrScanDialogSpinner",
  ],
  [
    "qrScanDialogSpinner",
  ],
]
`;

exports[`QrScanDialogComponent should test the scanning and component rendering correctly 3`] = `
[
  [
    {
      "message": "Kamera konnte gestartet werden",
    },
  ],
]
`;

exports[`QrScanDialogComponent should test the scanning and component rendering correctly 4`] = `
[
  [
    "/url",
  ],
]
`;

exports[`QrScanDialogComponent should test the scanning and component rendering correctly 5`] = `
<app-qr-scan-dialog
  action={[Function _NgxScannerQrcodeComponent]}
  componentName={[Function String]}
  data={[Function Object]}
  dialog={[Function _MatDialog]}
  obNotificationService={[Function _ObNotificationService]}
  spinnerChannel={[Function String]}
  spinnerService={[Function _ObSpinnerService]}
  subSink={[Function SubSink]}
>
  <div
    class="qr-scan-dialog__header"
    mat-dialog-content=""
  >
    <h2
      class="qr-scan-dialog__title"
      mat-dialog-title=""
    >
      title
    </h2>
  </div><ob-spinner
    class="ob-spinner"
  >
    <div
      class="ob-overlay ng-trigger ng-trigger-inOut"
    >
      <div
        class="ob-spinner-viewport"
      >
        <svg
          viewBox="0 0 48 48"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="24"
            cy="24"
            fill="none"
            r="20"
            stroke="#2F4356"
            stroke-miterlimit="10"
            stroke-width="4"
          />
        </svg>
      </div>
    </div>
  </ob-spinner><div
    class="qr-scan-dialog__content"
    mat-dialog-content=""
  >
    <ngx-scanner-qrcode
      class="ngx-scanner-qrcode"
    >
      <div
        class="origin-overlay"
      />
      <canvas
        class="origin-canvas"
      />
      <video
        class="origin-video"
        playsinline=""
      />
    </ngx-scanner-qrcode>
  </div><div
    class="qr-scan-dialog__actions"
    mat-dialog-actions=""
  >
    <button
      mat-button=""
      mat-dialog-close=""
      obbutton="primary"
      type="button"
    >
       i18n.common.close 
    </button>
  </div>
</app-qr-scan-dialog>
`;
